### ISSUE TITLE
[Bug Report]ReceiveAsync does not stop when passed cancellation token cancels

### ISSUE THREAD

**Issue by sensslen** [Azure/azure-iot-sdk-csharp#2246]
# Context

- **OS, version, SKU and CPU architecture used:** Windows 10 Desktop x64
- **Application's .NET Target Framework :** net6.0 
- **Device:** Laptop
- **SDK version used:** 1.39.0

## Description of the issue
When using Microsoft.Azure.Devices.Client.DeviceClient, I call ReceiveAsync to recieve messages from the hub. This works well until I want to gracefully shut down my application and thus cancel the cancellation token. This token is also passed to the ReceiveAsync method. Unfortunately the cancellation is not handled for around 30 seconds. 
_Above issue is resolved when using the latest main branch_

At this point I receive a IotHubCommunicationException (where I expect a OperationCanceledException as the documentation states).
_The issue about the wron exception is still present_

## Code sample exhibiting the issue
```csharp
using System.Text;
using Microsoft.Azure.Devices.Client;

var cancellationTokenSource = new CancellationTokenSource();
using var exitEvent = new AutoResetEvent(false);

Console.CancelKeyPress += (sender, args) =>
{
    Console.WriteLine("Request cancellation");

    cancellationTokenSource.Cancel();
    exitEvent.WaitOne();
    args.Cancel = true;
};

var url = "url";
var deviceId = "deviceId";
var symmetricKey = "key";
var authentication = new DeviceAuthenticationWithRegistrySymmetricKey(deviceId, symmetricKey);
using var client = DeviceClient.Create(url, authentication, TransportType.Amqp);
client.SetRetryPolicy(new ExponentialBackoff(10, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(10),
    TimeSpan.FromSeconds(2)));

var receptionTask = Reception(client, cancellationTokenSource.Token);

receptionTask.Wait();
Console.WriteLine("finished");
exitEvent.Set();

async Task Reception(DeviceClient client, CancellationToken cancellationToken)
{
    while (!cancellationToken.IsCancellationRequested)
    {
        try
        {
            using var message = await client.ReceiveAsync(cancellationToken);
            var messageBytes = message.GetBytes();
            Console.WriteLine($"Received:{Encoding.UTF8.GetString(messageBytes)}");
            await client.CompleteAsync(message, cancellationToken);
        }
        catch (OperationCanceledException)
        {
            // exepcted as of documenation
        }
        catch (Exception e)
        {
            // unexpected error
            Console.WriteLine(e);
        }
    }
}
```

## Console log of the issue
Please run the minimal code example

**Comment by abhipsaMisra** [comment]
This is due to the limitation of our amqp library dependency, which did not support cooperative cancellation but accepted a timeout instead. However, the amqp library has recently added cancellation token support, and we've incorporated the same into our SDK: https://github.com/Azure/azure-iot-sdk-csharp/pull/2216.
While we put an official release out, you are free to try out the code from `main` in the meanwhile and verify that it solves your issue.

FYI - Adding @azabbasi who implemented the amqp cancellation token support.

**Comment by azabbasi** [comment]
Thanks for reaching out to us regarding your issue.

The AMQP library has since fixed the issue and we have updated our SDK to respect these cancellation tokens. This change has not been published yet but it will be released with our next SDK version release.
You can see the PR here: #2216

**Comment by sensslen** [comment]
Thanks for the feedback! I will try Main then. Do you have a schedule for the next release?

**Comment by sensslen** [comment]
After using the code in the main branch of this repository, cancellation now works. However, I still get the "wrong" exception type thrown. I opened a ticket in the documentation (see Azure/azure-sdk-for-net#25643) which got closed in favour of this ticket. Should I open a new ticket concerning this exception or is this handled here?

**Comment by azabbasi** [comment]
I can take care of that one here, I will look into it right away.

**Comment by sensslen** [comment]
I actually don‘t mind the exception that is thrown. What I would appreciate is that the documentation matches the actual exceptions that are thrown. If there is more than one case where exceptions are thrown, I would really appreciate if this was documented. So to be clear: this is not an issue of the code but of it‘s documentation! This is why I filed the issue in the documentation - where it was closed because of this ticket!
